lib/Dialect/FIRRTL/Transforms/LowerIntrinsics.cpp

```diff
@@ -0,0 +1,171 @@
+//===- LowerIntrinsics.cpp - Lower Intrinsics -------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the LowerIntrinsics pass.  This pass processes FIRRTL
+// extmodules with intrinsic annotations and rewrites the instances as
+// appropriate.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLInstanceGraph.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/FIRRTL/FIRRTLTypes.h"
+#include "circt/Dialect/FIRRTL/FIRRTLVisitors.h"
+#include "circt/Dialect/FIRRTL/Namespace.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "mlir/IR/Diagnostics.h"
+#include "llvm/ADT/APSInt.h"
+#include "llvm/ADT/PostOrderIterator.h"
+#include "llvm/ADT/StringExtras.h"
+#include "llvm/Support/Debug.h"
+
+using namespace circt;
+using namespace firrtl;
+
+// Pass Infrastructure
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct LowerIntrinsicsPass : public LowerIntrinsicsBase<LowerIntrinsicsPass> {
+  void runOnOperation() override;
+};
+} // end anonymous namespace
+
+static bool lowerCirctSizeof(InstancePathCache& instancePathCache, FExtModuleOp mod) {
+      auto ports = mod.getPorts();
+      if (ports.size() != 2) {
```

Comment:
I wish there was a better way to separate verification from transformation, but I can't think of a better way for the short term implementation we have here...

Reply:
There's some table-driven verification possible.

Reply:
And we can fake it with some factoring.